File: misc/advent/2021/04/a.py
import math
from utils import flatten, printmat


def parse(f):
    ns = list(map(float, next(f).strip().split(",")))
    boards = []
    for line in f:
        line = line.strip()
        if not line:
            boards.append([])
            continue
        boards[-1].append(list(map(float, line.split())))
    return ns, boards[:-1]


def markboards(n, boards):
    for board in boards:
        for row in board:
            for i in range(len(board)):
                if row[i] == n:
                    row[i] = -1


def winner(board):
    l = len(board)
    for r in range(l):
        if all(board[r][i] < 0 for i in range(l)):
            return True
    for c in range(l):
        if all(board[i][c] < 0 for i in range(l)):
            return True


def bingo(ns, boards):
    l = len(boards[0][0])
    for n in ns:
        markboards(n, boards)

        for board in boards:
            if winner(board):
                return n, board


def score(n, board):
    return sum(x for x in flatten(board) if x >= 0) * n


print(score(*bingo(*parse(open("small.txt")))))
print(score(*bingo(*parse(open("input.txt")))))


def badbingo(ns, boards):
    l = len(boards[0][0])
    for n in ns:
        markboards(n, boards)

        liveboards = []
        for board in boards:
            if winner(board):
                if len(boards) == 1:
                    return n, board
                continue
            liveboards.append(board)

        boards = liveboards


print(score(*badbingo(*parse(open("small.txt")))))
print(score(*badbingo(*parse(open("input.txt")))))

File: misc/advent/2021/04/a.zig
// incomplete zig answer for day 4, trying to remember how to do stuff
const std = @import("std");

const Board = [25]i32;

const Bingo = struct {
    numbers: std.ArrayList(i32),
    boards: std.ArrayList(Board),
};

const OUT = std.io.getStdOut().writer();

// stole a bunch from https://zigbin.io/ad5582
fn parse(input: []const u8, alloc: std.mem.Allocator) !Bingo {
    var bingo = Bingo{
        .numbers = std.ArrayList(i32).init(alloc),
        .boards = std.ArrayList(Board).init(alloc),
    };
    var trimmed = std.mem.trim(u8, input, "\n");
    var data = std.mem.split(u8, trimmed, "\n\n");
    var ns = std.mem.split(u8, data.next().?, ",");
    while (ns.next()) |n| {
        try bingo.numbers.append(try std.fmt.parseInt(i32, n, 10));
    }
    while (data.next()) |board_data| {
        var board = try bingo.boards.addOne();
        var bns = std.mem.tokenize(u8, board_data, &std.ascii.spaces);
        var boardi: usize = 0;
        while (bns.next()) |bn| : (boardi += 1) {
            board[boardi] = try std.fmt.parseInt(i32, bn, 10);
        }
        std.debug.assert(boardi == 25);
    }

    return bingo;
}

fn run(bingo: Bingo) !void {
    for (bingo.numbers.items) |n| {
        try OUT.print("{d}\n", .{n});
    }
}

pub fn main() !void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();

    const allocator = arena.allocator();

    const args = try std.process.argsAlloc(allocator);
    const input = try std.fs.cwd().readFileAlloc(allocator, args[1], std.math.maxInt(usize));
    defer allocator.free(input);

    const bingo = try parse(input, allocator);
    try run(bingo);
}
